{"cells":[{"cell_type":"code","execution_count":17,"id":"78528b0e-2f01-4c35-abe1-897863d3498b","metadata":{"id":"78528b0e-2f01-4c35-abe1-897863d3498b","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1749580185436,"user_tz":240,"elapsed":385,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}},"outputId":"ddc83991-aa12-4a4c-d345-481129219395"},"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}],"source":["from google.colab import drive\n","drive.mount('/content/drive')"]},{"cell_type":"markdown","source":["# **1.\tLoad data**\n","\n","Load the given labeled dataset such that, each sentence from all documents is an element in the list.\n","\n","Load the given unlabeled dataset such that, each sentence from all documents is an element in the list.\n"],"metadata":{"id":"OaZRoXqz3XyU"},"id":"OaZRoXqz3XyU"},{"cell_type":"code","source":["import os\n","\n","labeled_path = \"/content/drive/MyDrive/NLP/Assigment5/labeled_datasets\"\n","unlabeled_path = \"/content/drive/MyDrive/NLP/Assigment5/unlabeled_datasets\"\n","\n","# Use the os library to read and combine all lines from each file in the labeled dataset directory\n","labeled_corpus = []\n","for filename in os.listdir(labeled_path):\n","    with open(os.path.join(labeled_path, filename), 'r') as f:\n","        lines = f.readlines()\n","        labeled_corpus.extend(lines)\n","\n","\n","unlabeled_corpus = []\n","for filename in os.listdir(unlabeled_path):\n","    with open(os.path.join(unlabeled_path, filename), 'r') as f:\n","        lines = f.readlines()\n","        unlabeled_corpus.extend(lines)\n","\n","print(labeled_corpus[:5])\n"],"metadata":{"id":"mkutrfG4Jihw","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1749580187087,"user_tz":240,"elapsed":34,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}},"outputId":"0c148251-426f-4adf-9c39-7fa71b264344"},"id":"mkutrfG4Jihw","execution_count":18,"outputs":[{"output_type":"stream","name":"stdout","text":["['### abstract ###\\n', 'AIMX\\tin this paper we derive the equations for loop corrected belief propagation on a continuous variable gaussian model\\n', 'OWNX\\tusing the exactness of the averages for belief propagation for gaussian models  a  different way of obtaining the covariances is found   based on belief propagation on cavity graphs\\n', 'OWNX\\twe discuss the relation of this  loop correction algorithm to expectation propagation  algorithms for the case in which the model is no longer  gaussian  but slightly perturbed by nonlinear terms\\n', '### introduction ###\\n']\n"]}]},{"cell_type":"markdown","source":["# **2.\tPre-process the text (on both labeled and unlabeled dataset)**\n","You can perform any of the followings:\n","•\tRemove the '### abstract ###' and '### introduction ###' titles\n","•\tRemove stopwords\n","•\tRemove punctuation\n","After this step, you should be working on a clean dataset.\n"],"metadata":{"id":"Ps0OLNNH4q7M"},"id":"Ps0OLNNH4q7M"},{"cell_type":"code","source":["!pip install nltk\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"l4yZRx2YNGmV","executionInfo":{"status":"ok","timestamp":1749580193496,"user_tz":240,"elapsed":3030,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}},"outputId":"129f9974-8974-42db-ef37-a8df47e7ad42"},"id":"l4yZRx2YNGmV","execution_count":19,"outputs":[{"output_type":"stream","name":"stdout","text":["Requirement already satisfied: nltk in /usr/local/lib/python3.11/dist-packages (3.9.1)\n","Requirement already satisfied: click in /usr/local/lib/python3.11/dist-packages (from nltk) (8.2.1)\n","Requirement already satisfied: joblib in /usr/local/lib/python3.11/dist-packages (from nltk) (1.5.1)\n","Requirement already satisfied: regex>=2021.8.3 in /usr/local/lib/python3.11/dist-packages (from nltk) (2024.11.6)\n","Requirement already satisfied: tqdm in /usr/local/lib/python3.11/dist-packages (from nltk) (4.67.1)\n"]}]},{"cell_type":"code","source":["import nltk\n","nltk.download('stopwords')\n","nltk.download('punkt_tab')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"TIEMltkNN3Sl","executionInfo":{"status":"ok","timestamp":1749580194397,"user_tz":240,"elapsed":17,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}},"outputId":"8a4027cf-67dd-4707-aec3-0d71e846482e"},"id":"TIEMltkNN3Sl","execution_count":20,"outputs":[{"output_type":"stream","name":"stderr","text":["[nltk_data] Downloading package stopwords to /root/nltk_data...\n","[nltk_data]   Package stopwords is already up-to-date!\n","[nltk_data] Downloading package punkt_tab to /root/nltk_data...\n","[nltk_data]   Package punkt_tab is already up-to-date!\n"]},{"output_type":"execute_result","data":{"text/plain":["True"]},"metadata":{},"execution_count":20}]},{"cell_type":"code","source":["import re\n","import string\n","from nltk.corpus import stopwords\n","from nltk.tokenize import word_tokenize\n","\n","#  Preprocess the text\n","nltk.download('stopwords')\n","stop_words = set(stopwords.words('english'))\n","\n","def preprocess(text):\n","    text = text.lower()\n","    text = re.sub(r\"^#{3}\\s?[a-zA-Z]+\\s?#{3}$\", \"\", text.strip())\n","    tokens = word_tokenize(text)\n","    tokens = [t for t in tokens if t not in string.punctuation]\n","    tokens = [t for t in tokens if t not in stop_words]\n","    return \" \".join(tokens)\n","\n","# Extract labels and preprocess each labeled sentence\n","labels = []\n","cleaned_labeled_sentences = []\n","\n","for line in labeled_corpus:\n","    if \"\\t\" in line:\n","        tag, sentence = line.strip().split(\"\\t\", 1) # split\n","        labels.append(tag) # store labels\n","        cleaned_labeled_sentences.append(preprocess(sentence)) # store sentence\n","\n","cleaned_unlabeled_sentences = [preprocess(s.strip()) for s in unlabeled_corpus]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ntovBMjcNBjF","executionInfo":{"status":"ok","timestamp":1749580203522,"user_tz":240,"elapsed":97,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}},"outputId":"ea1347af-a072-42a5-877c-0a03f556421c"},"id":"ntovBMjcNBjF","execution_count":22,"outputs":[{"output_type":"stream","name":"stderr","text":["[nltk_data] Downloading package stopwords to /root/nltk_data...\n","[nltk_data]   Package stopwords is already up-to-date!\n"]}]},{"cell_type":"markdown","source":["# **3.\tVectorize the clean labeled dataset**\n","\n"],"metadata":{"id":"_kkfHIlx5hhZ"},"id":"_kkfHIlx5hhZ"},{"cell_type":"code","source":["from sklearn.feature_extraction.text import TfidfVectorizer\n","\n","vectorizer = TfidfVectorizer()\n","vectorizer.fit(cleaned_labeled_sentences + cleaned_unlabeled_sentences)\n","\n","# label\n","x = vectorizer.transform(cleaned_labeled_sentences)\n","y = labels\n","# unlabels\n","x_unlabeled = vectorizer.transform(cleaned_unlabeled_sentences)"],"metadata":{"id":"mXIhjoTqRfRR","executionInfo":{"status":"ok","timestamp":1749580206904,"user_tz":240,"elapsed":58,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}}},"id":"mXIhjoTqRfRR","execution_count":23,"outputs":[]},{"cell_type":"markdown","source":["# **4.\tCreate the Logistic Regression classifier (sample code available in blackboard)**"],"metadata":{"id":"wKIx_S5k6Dr1"},"id":"wKIx_S5k6Dr1"},{"cell_type":"code","source":["from sklearn.model_selection import train_test_split\n","from sklearn.linear_model import LogisticRegression\n","\n","x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)\n","\n","log_reg = LogisticRegression()\n","log_reg.fit(x_train, y_train)\n","y_pred_log = log_reg.predict(x_test)"],"metadata":{"id":"HY_X4TDrRfOE","executionInfo":{"status":"ok","timestamp":1749580209908,"user_tz":240,"elapsed":147,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}}},"id":"HY_X4TDrRfOE","execution_count":24,"outputs":[]},{"cell_type":"markdown","source":["# **5.\tEvaluate the Logistic Regression classifier**"],"metadata":{"id":"tp--vIMM8AUy"},"id":"tp--vIMM8AUy"},{"cell_type":"code","source":["from sklearn.metrics import classification_report\n","print(\"\\n=== Logistic Regression Report ===\")\n","log_reg_report = classification_report(\n","    y_test, y_pred_log,\n","    labels = ['AIMX', 'OWNX', 'CONT', 'BASE', 'NUMBER', 'MISC'],\n","    zero_division=0 # to avoid warning\n",")\n","print(log_reg_report)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"wfEGv-63RfA1","executionInfo":{"status":"ok","timestamp":1749580329174,"user_tz":240,"elapsed":108,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}},"outputId":"067a1276-0ae3-4480-8ab7-b977ecb76fe7"},"id":"wfEGv-63RfA1","execution_count":29,"outputs":[{"output_type":"stream","name":"stdout","text":["\n","=== Logistic Regression Report ===\n","              precision    recall  f1-score   support\n","\n","        AIMX       1.00      0.50      0.67         2\n","        OWNX       0.43      0.67      0.52         9\n","        CONT       1.00      0.33      0.50         6\n","        BASE       0.00      0.00      0.00         1\n","      NUMBER       0.00      0.00      0.00         0\n","        MISC       0.71      0.75      0.73        16\n","\n","    accuracy                           0.62        34\n","   macro avg       0.52      0.38      0.40        34\n","weighted avg       0.68      0.62      0.61        34\n","\n"]}]},{"cell_type":"markdown","source":["# **6.\tTry other classifiers: support vector machine and decision tree. **"],"metadata":{"id":"KiXUe1-48VOI"},"id":"KiXUe1-48VOI"},{"cell_type":"code","source":["from sklearn.svm import SVC\n","from sklearn.tree import DecisionTreeClassifier\n","from sklearn.metrics import classification_report\n","from sklearn.naive_bayes import MultinomialNB\n","\n","# SVM\n","svm = SVC()\n","svm.fit(x_train, y_train)\n","y_pred_svm = svm.predict(x_test)\n","print(\"\\n=== SVM Report ===\")\n","print(classification_report(\n","    y_test, y_pred_svm,\n","    labels=['AIMX', 'OWNX', 'CONT', 'BASE', 'NUMBER', 'MISC'],\n","    zero_division=0\n","))\n","\n","# Decision Tree\n","tree = DecisionTreeClassifier()\n","tree.fit(x_train, y_train)\n","y_pred_tree = tree.predict(x_test)\n","print(\"\\n=== Decision Tree Report ===\")\n","print(classification_report(\n","    y_test, y_pred_tree,\n","    labels=['AIMX', 'OWNX', 'CONT', 'BASE', 'NUMBER', 'MISC'],\n","    zero_division=0\n","))\n","\n","# NaiveBayes\n","NaiveBayes= MultinomialNB()\n","NaiveBayes.fit(x_train, y_train)\n","y_pred_NB = NaiveBayes.predict(x_test)\n","print(\"\\n=== Naive Bayes Report ===\")\n","print(classification_report(\n","    y_test, y_pred_NB,\n","    labels=['AIMX', 'OWNX', 'CONT', 'BASE', 'NUMBER', 'MISC'],\n","    zero_division=0\n","))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"6uGj7xCdSWcF","executionInfo":{"status":"ok","timestamp":1749580331535,"user_tz":240,"elapsed":198,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}},"outputId":"6ee51e68-49a1-433b-d6b0-0e3ff853f313"},"id":"6uGj7xCdSWcF","execution_count":30,"outputs":[{"output_type":"stream","name":"stdout","text":["\n","=== SVM Report ===\n","              precision    recall  f1-score   support\n","\n","        AIMX       1.00      0.50      0.67         2\n","        OWNX       0.50      1.00      0.67         9\n","        CONT       1.00      0.33      0.50         6\n","        BASE       0.00      0.00      0.00         1\n","      NUMBER       0.00      0.00      0.00         0\n","        MISC       0.85      0.69      0.76        16\n","\n","    accuracy                           0.68        34\n","   macro avg       0.56      0.42      0.43        34\n","weighted avg       0.77      0.68      0.66        34\n","\n","\n","=== Decision Tree Report ===\n","              precision    recall  f1-score   support\n","\n","        AIMX       1.00      0.50      0.67         2\n","        OWNX       0.40      0.89      0.55         9\n","        CONT       1.00      0.50      0.67         6\n","        BASE       0.00      0.00      0.00         1\n","      NUMBER       0.00      0.00      0.00         0\n","        MISC       0.80      0.50      0.62        16\n","\n","    accuracy                           0.59        34\n","   macro avg       0.53      0.40      0.42        34\n","weighted avg       0.72      0.59      0.59        34\n","\n","\n","=== Naive Bayes Report ===\n","              precision    recall  f1-score   support\n","\n","        AIMX       0.00      0.00      0.00         2\n","        OWNX       0.42      0.89      0.57         9\n","        CONT       1.00      0.33      0.50         6\n","        BASE       0.00      0.00      0.00         1\n","      NUMBER       0.00      0.00      0.00         0\n","        MISC       0.77      0.62      0.69        16\n","\n","    accuracy                           0.59        34\n","   macro avg       0.37      0.31      0.29        34\n","weighted avg       0.65      0.59      0.56        34\n","\n"]}]},{"cell_type":"code","source":["# === Predict on Unlabeled Data ===\n","unlabeled_predictions = NaiveBayes.predict(x_unlabeled)\n","\n","print(\"\\n Printout of Predicted Labels for Unlabeled Sentences:\\n\")\n","for label, sentence in zip(unlabeled_predictions, unlabeled_corpus):\n","    print(f\"{label}\\t{sentence.strip()}\")\n"],"metadata":{"id":"uIanKmcLRT0l","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1749580334673,"user_tz":240,"elapsed":21,"user":{"displayName":"Kevin Nguyen","userId":"01694590780485184673"}},"outputId":"0b1156f2-514d-4dfc-9474-4c688b54784c"},"id":"uIanKmcLRT0l","execution_count":31,"outputs":[{"output_type":"stream","name":"stdout","text":["\n"," Printout of Predicted Labels for Unlabeled Sentences:\n","\n","MISC\t### abstract ###\n","OWNX\tWhole-genome transporter analyses have been conducted on 141 organisms whose complete genome sequences are available.\n","OWNX\tFor each organism, the complete set of membrane transport systems was identified with predicted functions, and classified into protein families based on the transporter classification system.\n","OWNX\tOrganisms with larger genome sizes generally possessed a relatively greater number of transport systems.\n","MISC\tIn prokaryotes and unicellular eukaryotes, the significant factor in the increase in transporter content with genome size was a greater diversity of transporter types.\n","OWNX\tIn contrast, in multicellular eukaryotes, greater number of paralogs in specific transporter families was the more important factor in the increase in transporter content with genome size.\n","OWNX\tBoth eukaryotic and prokaryotic intracellular pathogens and endosymbionts exhibited markedly limited transport capabilities.\n","OWNX\tHierarchical clustering of phylogenetic profiles of transporter families, derived from the presence or absence of a certain transporter family, showed that clustering patterns of organisms were correlated to both their evolutionary history and their overall physiology and lifestyles.\n","MISC\t### introduction ###\n","OWNX\tMembrane transport systems play essential roles in cellular metabolism and activities.\n","OWNX\tTransporters function in the acquisition of organic nutrients, maintenance of ion homeostasis, extrusion of toxic and waste compounds, environmental sensing and cell communication, and other important cellular functions CITATION.\n","MISC\tVarious transport systems differ in their putative membrane topology, energy coupling mechanisms, and substrate specificities CITATION.\n","OWNX\tAmong the prevailing energy sources are adenosine triphosphate, phosphoenolpyruvate, and chemiosmotic energy in the form of sodium ion or proton electrochemical gradients.\n","OWNX\tThe transporter classification system represents a systematic approach to classify transport systems according to their mode of transport, energy coupling mechanism, molecular phylogeny, and substrate specificity CITATION CITATION.\n","OWNX\tTransport mode and energy coupling mechanism serve as the primary basis for classification because of their relatively stable characteristics.\n","OWNX\tThere are four major classes of solute transporters in the transporter classification system: channels, primary transporters, secondary transporters, and group translocators.\n","OWNX\tTransporters of unknown mechanism or function are included as a distinct class.\n","MISC\tChannels are energy-independent transporters that transport water, specific types of ions, or hydrophilic small molecules down a concentration or electrical gradient; they have higher rates of transport and lower stereospecificity than the other transporter classes.\n","OWNX\tPrimary active transporters couple the transport process to a primary source of energy.\n","OWNX\tSecondary transporters utilize an ion or solute electrochemical gradient, e.g., proton/sodium motive force, to drive the transport process.\n","MISC\tE. coli LacY lactose permease CITATION, CITATION is probably one of the best characterized secondary transporters CITATION.\n","OWNX\tGroup translocators modify their substrates during the transport process.\n","OWNX\tFor example, E. coli MtlA mannitol PTS transporter phosphorylates exogenous mannitol using phosphoenolpyruvate as the phosphoryl donor and energy source and releases the phosphate ester, mannitol-1-P, into the cell cytoplasm CITATION, CITATION.\n","OWNX\tEach transporter class is further classified into individual families and subfamilies according to their function, phylogeny, and/or substrate specificity CITATION .\n","MISC\tSince the advent of genomic sequencing technologies, the complete sequences of over 200 prokaryotic and eukaryotic genomes have been published to date, representing a wide range of species from archaea to human.\n","OWNX\tThere are also more than 1,100 additional genome sequencing projects currently underway around the world CITATION, CITATION.\n","MISC\tConvenient and effective computational methods are required to handle and analyze the immense amount of data generated by the whole-genome sequencing projects.\n","OWNX\tAn in-depth look at transport proteins is vital to the understanding of the metabolic capability of sequenced organisms.\n","MISC\tHowever, it is often problematic to annotate these transport proteins by current primary annotation methods because of the occurrence of large and complex transporter gene families, such as the ATP-binding cassette superfamily CITATION, CITATION and the major facilitator superfamily CITATION, CITATION, and the presence of multiple transporter gene paralogs in many organisms.\n","MISC\tWe have been working on a systematic genome-wide analysis of cellular membrane transport systems.\n","MISC\tPreviously, we reported a comprehensive analysis of the transport systems in 18 prokaryotic organisms CITATION, CITATION and in yeast CITATION.\n","OWNX\tHere we expand our analyses to 141 species and compare the fundamental differences in membrane transport systems in prokaryotes and eukaryotes.\n","OWNX\tPhylogenetic profiling of transporter families and predicted substrates was utilized to investigate the relevance of transport capabilities to the overall physiology of prokaryotes and eukaryotes.\n","MISC\t### abstract ###\n","MISC\tA central problem in the bioinformatics of gene regulation is to find the binding sites for regulatory proteins.\n","MISC\tOne of the most promising approaches toward identifying these short and fuzzy sequence patterns is the comparative analysis of orthologous intergenic regions of related species.\n","MISC\tThis analysis is complicated by various factors.\n","MISC\tFirst, one needs to take the phylogenetic relationship between the species into account in order to distinguish conservation that is due to the occurrence of functional sites from spurious conservation that is due to evolutionary proximity.\n","MISC\tSecond, one has to deal with the complexities of multiple alignments of orthologous intergenic regions, and one has to consider the possibility that functional sites may occur outside of conserved segments.\n","OWNX\tHere we present a new motif sampling algorithm, PhyloGibbs, that runs on arbitrary collections of multiple local sequence alignments of orthologous sequences.\n","OWNX\tThe algorithm searches over all ways in which an arbitrary number of binding sites for an arbitrary number of transcription factors can be assigned to the multiple sequence alignments.\n","MISC\tThese binding site configurations are scored by a Bayesian probabilistic model that treats aligned sequences by a model for the evolution of binding sites and background intergenic DNA.\n","MISC\tThis model takes the phylogenetic relationship between the species in the alignment explicitly into account.\n","OWNX\tThe algorithm uses simulated annealing and Monte Carlo Markov-chain sampling to rigorously assign posterior probabilities to all the binding sites that it reports.\n","OWNX\tIn tests on synthetic data and real data from five Saccharomyces species our algorithm performs significantly better than four other motif-finding algorithms, including algorithms that also take phylogeny into account.\n","OWNX\tOur results also show that, in contrast to the other algorithms, PhyloGibbs can make realistic estimates of the reliability of its predictions.\n","OWNX\tOur tests suggest that, running on the five-species multiple alignment of a single gene's upstream region, PhyloGibbs on average recovers over 50 percent of all binding sites in S. cerevisiae at a specificity of about 50 percent, and 33 percent of all binding sites at a specificity of about 85 percent.\n","OWNX\tWe also tested PhyloGibbs on collections of multiple alignments of intergenic regions that were recently annotated, based on ChIP-on-chip data, to contain binding sites for the same TF.\n","OWNX\tWe compared PhyloGibbs's results with the previous analysis of these data using six other motif-finding algorithms.\n","MISC\tFor 16 of 21 TFs for which all other motif-finding methods failed to find a significant motif, PhyloGibbs did recover a motif that matches the literature consensus.\n","OWNX\tIn 11 cases where there was disagreement in the results we compiled lists of known target genes from the literature, and found that running PhyloGibbs on their regulatory regions yielded a binding motif matching the literature consensus in all but one of the cases.\n","OWNX\tInterestingly, these literature gene lists had little overlap with the targets annotated based on the ChIP-on-chip data.\n","OWNX\tThe PhyloGibbs code can be downloaded from LINK or LINK.\n","OWNX\tThe full set of predicted sites from our tests on yeast are available at LINK.\n","MISC\t### introduction ###\n","MISC\tTranscription factors are proteins that bind in a sequence-specific manner to short DNA segments, most commonly in intergenic DNA upstream of a gene, to activate or suppress gene transcription.\n","OWNX\tTheir DNA-binding domains recognize collections of short related DNA sequences.\n","OWNX\tOne generally finds that, although there is no unique combination of bases that is shared by all binding sites, and although different bases can occur at each position, there are clear biases in the distribution of bases that occur at each position of the binding sites.\n","MISC\tA common mathematical representation of a motif that takes this variability into account is a so-called weight matrix CITATION, CITATION w, whose components w i give the probabilities of finding base A, C, G, T at position i of a binding site.\n","MISC\tThe main assumption underlying this mathematical representation is that the bases occurring at different positions of the binding site are probabilistically independent.\n","OWNX\tThis in turn follows, under some conditions CITATION, from the assumption that the binding energy of the protein to the DNA is a sum of pairwise contact energies between the individual nucleotides and the protein.\n","OWNX\tThere are several algorithms that are based on the WM representation that detect, ab initio, binding sites for a common TF in a collection of DNA sequences CITATION CITATION.\n","OWNX\tThese algorithms broadly fall into two classes.\n","OWNX\tOne class, of which MEME CITATION is the typical representative, searches the space of all WMs for the WM that can best explain the observed sequences.\n","OWNX\tThe class of Gibbs sampling algorithms, of which the Gibbs motif sampler CITATION, CITATION is the typical representative, instead samples the space of all multiple alignments of small sequence segments in search of the one that is most likely to consist of samples from a common WM.\n","OWNX\tA crucial factor for the success of ab initio methods is the ratio of the number of binding sites to the total amount of DNA in the collection of sequences.\n","OWNX\tThat is, the larger the number of binding sites in the set, and the smaller the total amount of DNA, the more likely it is that ab initio methods can discover the binding sites among the other DNA sequences.\n","MISC\tIn order to ensure a reasonable chance of success one thus needs to provide these methods with collections of sequences that are highly enriched with binding sites for a common TF.\n","OWNX\tOne possibility is to use sets of upstream regions from genes that appear co-regulated in microarray experiments or that were bound by a common TF in ChIP-on-chip experiments.\n","OWNX\tAnother possibility is to use upstream regions of orthologous genes from related organisms.\n","MISC\tHere the assumption is that the regulation of the ancestor gene, and thus its binding sites, has been conserved in the orthologs that descend from it.\n","OWNX\tThis latter approach is in general complicated by a number of factors.\n","MISC\tWhen searching for regulatory sites in sequences that are not phylogenetically related, such as upstream regions of different genes from the same organism, one may simply look for short sequence motifs that are overrepresented among the input sequences.\n","OWNX\tIf the set of species from which the orthologous sequences derive are sufficiently diverged, one may simply choose to ignore the phylogenetic relationship between the sequences and treat the orthologous sequences in the same way as sequences that are not phylogenetically related.\n","OWNX\tThis was, for instance, the approach taken by McCue et al. CITATION, CITATION, where the Gibbs motif sampler algorithm CITATION, CITATION was used on upstream regions of proteo- bacteria.\n","OWNX\tHowever, this approach is not applicable to datasets containing more closely related species, where some of the sequences will exhibit significant amounts of similarity simply because of their evolutionary proximity.\n","OWNX\tMoreover, the amount of similarity will depend on the phylogenetic distance between the species, and it is clear that finding conserved sequence motifs between orthologous sequences from closely related species is much less indicative of function than finding sequence motifs that are conserved between distant species.\n","MISC\tOne will in general thus have to distinguish conservation due to functional constraints from conservation due to evolutionary proximity, and to do this correctly, the phylogenetic relationship between the sequences has to be taken into account.\n","OWNX\tA second challenge in using orthologous intergenic sequences from multiple species is the nontrivial structure of their multiple alignments.\n","MISC\tOne typically finds a very heterogeneous pattern of conservation: well-conserved blocks of different sizes and covering different subsets of the species are interspersed with sequence segments that show little similarity with the sequences of the other species.\n","OWNX\tThe technique of phylogenetic footprinting, restricts attention to only those sequence segments in the genome of interest that show significant conservation with the other species.\n","MISC\tThe conserved regions for multiple genes are then searched for common motifs by a variety of techniques.\n","MISC\tIt is unclear, however, to what extent regulatory sites are restricted to such conserved segments.\n","MISC\tFor instance, several studies of Drosophila and yeast CITATION CITATION have shown that there is no strong correlation between where experimentally annotated binding sites occur and whether that region is conserved.\n","OWNX\tThus, at least for yeast and flies, considerable information is lost by focusing on the conserved regions only.\n","MISC\tWe thus decided to retain the entire patchwork pattern of conserved sequence blocks and unaligned segments.\n","OWNX\tOur strategy is implemented by a Gibbs sampling approach, and a preliminary account of the algorithm was presented in CITATION.\n","OWNX\tThe algorithm operates on arbitrary collections of both phylogenetically related sequences, such as orthologous intergenic regions, and sequences that are not phylogenetically related, such as upstream regions of different genes from the same organism.\n","OWNX\tThe phylogenetically related groups of sequences in the input are pre-aligned into local multiple alignments where clearly similar sequence segments are aligned into blocks and sequence segments of no or marginal similarity are left unaligned CITATION.\n","OWNX\tAlthough the algorithm can also take global multiple alignments as input, we believe that these often force phylogenetically unrelated segments into aligned blocks.\n","OWNX\tThis may adversely affect the performance of the algorithm.\n","OWNX\tWe score putative sites within blocks of aligned sequences with an evolutionary model that takes the phylogenetic relationships of the species into account, while putative sites in unaligned segments are treated as independent occurrences.\n","OWNX\tThis Bayesian model defines a probability distribution over arbitrary placements of putative binding sites for multiple motifs, and we sample it with a Monte Carlo Markov chain.\n","OWNX\tWe first use simulated annealing to search for the globally optimal configuration of binding sites.\n","MISC\tThe motifs in this configuration are then tracked in a further sampling run to estimate realistic posterior probabilities for all the binding sites that the algorithm reports.\n","OWNX\tRecently a number of other algorithms have been developed that search for regulatory motifs in groups of phylogenetically related sequences.\n","OWNX\tProbably the first algorithm that was proposed is a generalization of the Consensus algorithm CITATION called PhyloCon CITATION.\n","OWNX\tPhyloCon operates on sets of co-regulated genes and their orthologs.\n","OWNX\tIt is a greedy algorithm that first finds ungapped alignments of similar sequence segments in sets of orthologous sequences, and then combines these alignments from different upstream regions into larger alignments.\n","OWNX\tThis algorithm does not take any phylogenetic information into account, i.e., closely related sequences are treated the same as distantly related sequences.\n","OWNX\tOther drawbacks of this algorithm are that it assumes that each motif will have exactly one site in each of the intergenic regions and that it assumes that this site is conserved in all orthologs.\n","OWNX\tMore closely related to PhyloGibbs's approach are two recent algorithms CITATION, CITATION that generalize MEME CITATION to take the phylogenetic relationships between species into account.\n","OWNX\tThe main difference between EMnEM and PhyME is that PhyME uses the same evolutionary model for the evolution of binding sites as PhyloGibbs, which takes into account that binding sites evolve under constraints set by a WM, whereas EMnEM simply assumes an overall slower rate of evolution in binding sites than in background sequences.\n","MISC\tAnother difference is that PhyME, like PhyloGibbs, treats the multiple alignment more flexibly than EMnEM, which demands a global multiple alignment.\n","OWNX\tThe main difference between PhyloGibbs and these algorithms is of course that PhyloGibbs takes a motif sampling approach, which allows us to search for multiple motifs in parallel, whereas PhyME and EMnEM use expectation maximization to search for one WM at a time.\n","OWNX\tIn the following sections, we first describe our Bayesian model that assigns a posterior probability to each configuration of binding sites for multiple motifs assigned to the input sequences.\n","OWNX\tWe start by describing the model for phylogenetically unrelated sequences, which is essentially equivalent to the model used in the Gibbs motif sampler CITATION, CITATION, and then describe how this model is extended to datasets that contain phylogenetically related sequences.\n","OWNX\tAfter that we describe the move set with which we search the state space of all possible configurations, and the annealing and tracking strategy that we use to identify the significant groups of sites.\n","OWNX\tWe then present examples of the performance of ours and other algorithms on both synthetic and real data.\n","OWNX\tThe synthetic datasets consist of mixtures of WM samples and random sequences, which is in accordance with assumptions that all algorithms make.\n","OWNX\tThis allows us to compare the performance of the algorithms in an idealized situation that does not contain the complexities of real data.\n","OWNX\tThese tests also show to what extent binding sites can be recovered for this idealized data as a function of the quality of the WMs, the number of sites available, and the number of species available and their phylogenetic distances.\n","OWNX\tFor our tests on real data we use 200 upstream regions from Saccharomyces cerevisiae that have known binding sites from the collection CITATION, and compare the ability of the different algorithms to recover these sites when running on multiple alignments of the orthologs of these upstream regions from recently sequenced Saccharomyces genomes CITATION, CITATION.\n","OWNX\tFinally, we run PhyloGibbs on collections of upstream region alignments that were annotated in CITATION to contain binding sites for a common TF based on data from ChIP-on-chip experiments, and we extensively compare PhyloGibbs' results with the annotations in CITATION and with the literature.\n"]}]},{"cell_type":"markdown","source":["# **Summary **\n","In this project, I worked with two sets of datasets: one labeled and one unlabeled. I first cleaned both datasets by removing headers, stopwords, and punctuation. Then I used TF-IDF to turn the sentences into numeric features.\n","\n","I trained a Logistic Regression model to classify the sentences into six categories (AIMX, OWNX, CONT, BASE, NUMBER, MISC) using the labeled data. I tested the model's performance using accuracy reports. I also compared it with other models like SVM, Decision Tree, and Naive Bayes.\n","\n","Finally, I used the trained Logistic Regression model to predict the categories for the unlabeled sentences and printed the predicted labels along with each sentence.\n","\n","This process showed that Logistic Regression worked well for this classification task, and the model was able to apply the learned categories to unseen data."],"metadata":{"id":"i1TvxQm0B-Rc"},"id":"i1TvxQm0B-Rc"}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.9"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}